### PR TITLE
[SimpleFSDP] Add tensor parallel support

### PR DESCRIPTION
**Pull request by ruisizhang123**
This PR adds tensor parallel (TP) support for SimpleFSDP, together with [this pr](https://github.com/pytorch/pytorch/pull/152286) from PyTorch.

**(This is a placeholder for now, and shall be merged only after the PyTorch pr is landed.)**

**Profile Trace & Convergence on debug model:**

Screenshot 2025-04-27 at 3 48 35 PM

Screenshot 2025-04-27 at 3 31 46 PM

The loss curves are all benchmarked with `training.seed = 42` on 4 GPUs.

The end-to-end SimpleFSDP TP integration has been proved to work in the PR from this fork: https://github.com/tianyu-l/pytorch_intern24/pull/25. Per the discussion with Tianyu, the new PyTorch PR has add `_StridedShard` following FSDP2 to be compatible with distributed checkpointing. The newly benchmarked results demonstrated it works properly.

### CODE REVIEW

File: torchtitan/experiments/simple_fsdp/parallelize_llama.py [hunk 1]
```diff
@@ -77,6 +76,11 @@ def parallelize_llama(
             dp_mesh_dim_names = ("dp_shard_cp",)
             dp_mode = "fully_shard"
 
+        world_mesh.mesh_dim_names = (
+            tuple(dp_mesh_dim_names)
+            + world_mesh.mesh_dim_names[len(dp_mesh_dim_names) :]
+        )
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/parallelize_llama.py]: oh do we need this? we didn't have it in https://github.com/tianyu-l/torchtitan_intern24/blob/main/torchtitan/parallelisms/parallelize_llama.py

File: torchtitan/experiments/simple_fsdp/simple_fsdp.py [hunk 2]
```diff
@@ -44,6 +49,92 @@ class MixedPrecisionPolicy:
     reduce_dtype: Optional[torch.dtype] = None
 
 
+def _distribute_tensor(
+    tensor: torch.Tensor,
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/simple_fsdp.py]: type should be dtensor
In fact, since this function is specialized, we may call this function `_distribute_dtensor` and add comments saying this is only for applying FSDP sharding on top of TP sharded 1-D DTensor

File: torchtitan/experiments/simple_fsdp/simple_fsdp.py [hunk 3]
```diff
@@ -44,6 +49,92 @@ class MixedPrecisionPolicy:
...
+    device_mesh: Optional[DeviceMesh] = None,
+    placements: Optional[Sequence[Placement]] = None,
+) -> DTensor:
+    # get default device mesh if there's nothing specified
+    device_mesh = device_mesh or _mesh_resources.get_current_mesh()
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/simple_fsdp.py]: let's mark `device_mesh` as required and not do this dynamic fetching

File: torchtitan/experiments/simple_fsdp/simple_fsdp.py [hunk 4]
```diff
@@ -44,6 +49,92 @@ class MixedPrecisionPolicy:
...
+
+    # set default placements to replicated if not specified
+    if placements is None:
+        placements = [Replicate() for _ in range(device_mesh.ndim)]
+
+    if len(placements) != device_mesh.ndim:
+        raise ValueError(
+            f"`placements` must have the same length as `device_mesh.ndim`! "
+            f"Found placements length: {len(placements)}, and device_mesh.ndim: {device_mesh.ndim}."
+        )
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/simple_fsdp.py]: I think we can remove these general code in this specialized function

File: torchtitan/experiments/simple_fsdp/simple_fsdp.py [hunk 5]
```diff
@@ -44,6 +49,92 @@ class MixedPrecisionPolicy:
...
+
+    # NOTE: Below are experimental enhancements to distribute a DTensor.
+    #       This helps enable Simple FSDP + TP, in which
+    #       inner spec/mesh is TP spec/mesh
+    #       outer spec/mesh is FSDP spec/mesh
+    # The logic follows
+    # https://github.com/pytorch/pytorch/blob/main/torch/distributed/_composable/fsdp/_fsdp_param.py#L261
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/simple_fsdp.py]: Let's make it part of the comments / docstring of this function.

File: torchtitan/experiments/simple_fsdp/parallelize_llama.py (hunk 1 above)
Comment (ruisizhang123) [reply torchtitan/experiments/simple_fsdp/parallelize_llama.py]: yes, if we don't update the global`world_mesh.mesh_dim_names` here, when `x._spec.mesh` is trying to get a slice of `tp` mesh in [`replicate_compute`](https://github.com/pytorch/torchtitan/blob/a46fd349c9d074829adb42dc7b49d883f74f47de/torchtitan/experiments/simple_fsdp/simple_fsdp.py#L180), the code cannot pass this check [here](https://github.com/pytorch/pytorch/blob/5d36485b4aa5823bb9ee5eb070548c4ee355f3b2/torch/distributed/device_mesh.py#L298-L299).

The reason is `x._spec.mesh` is device mesh is updated as `mesh_dim_names=('dp_shard_cp', 'tp')`; however, the global `world_mesh.mesh_dim_names` is still `mesh_dim_names=('dp_shard', 'tp')`.

File: torchtitan/experiments/simple_fsdp/simple_fsdp.py [hunk 6]
```diff
@@ -44,6 +49,72 @@ class MixedPrecisionPolicy:
...
+) -> DTensor:
+    """
+    Below are experimental enhancements to distribute a DTensor.
+    This helps enable Simple FSDP + TP, in which
+        inner spec/mesh is TP spec/mesh
+        outer spec/mesh is FSDP spec/mesh
+    The logic follows
+    https://github.com/pytorch/pytorch/blob/main/torch/distributed/_composable/fsdp/_fsdp_param.py#L261
+    """
+    device_mesh = device_mesh or _mesh_resources.get_current_mesh()
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/simple_fsdp.py]: let's remove this line

File: torchtitan/experiments/simple_fsdp/parallelize_llama.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/experiments/simple_fsdp/parallelize_llama.py]: hmm I'm not sure if `world_mesh.mesh_dim_names` is supposed to be updated in this this way and if it has any unexpected side-effect. E.g. have you checked if it works with FSDP+TP+CP?

@fegin to share more

File: torchtitan/experiments/simple_fsdp/parallelize_llama.py [hunk 7]
```diff
@@ -77,6 +76,11 @@ def parallelize_llama(
             dp_mesh_dim_names = ("dp_shard_cp",)
             dp_mode = "fully_shard"
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/parallelize_llama.py]: When I worked on https://github.com/tianyu-l/pytorch_intern24/pull/25 I was aiming at FSDP+TP. Now it's a good time to make sure `hybrid_shard` and `replicate` modes also work with TP. Depending on the complexity, feel free to work in separate PRs.

File: torchtitan/experiments/simple_fsdp/parallelize_llama.py (hunk 1 above)
Comment (fegin) [reply torchtitan/experiments/simple_fsdp/parallelize_llama.py]: It's not supposed to be updated. But I can see why this is an issue for `SimpleFSDP` but not `fully_shard`. The reason is that `fully_shard` get a DTensor with TP submesh while `SimpleFSDP` gets a DP + TP submesh in which case you have to slice again. And unfortunately, the design of DeviceMesh intentionally not to support this case.

The side effect is that we won't be able to get `world_mesh["dp_shard", "dp_replicate"]` if we need.

Comment (tianyu-l) [reply torchtitan/experiments/simple_fsdp/parallelize_llama.py]: I just tested:
- FSDP + TP + CP will encounter the same error https://github.com/pytorch/pytorch/blob/5d36485b4aa5823bb9ee5eb070548c4ee355f3b2/torch/distributed/device_mesh.py#L298-L299
- CP + TP will fail because 'cp' mesh cannot be found by CP module due to `world_mesh.mesh_dim_names` change.

So let's don't go with this change.

In the short term, as a temp remedy, let's explicitly pass the `tp_mesh` into the `data_parallel` API as `self.tp_mesh` so it can be used by `replicate_compute` (`None` if TP not used). I just tested locally that it works well with FSDP and CP.

In the long term, I believe DeviceMesh should allow slicing a submesh. 
Let's keep the original code commented and add this as a TODO item. Once it's ready we can remove the `tp_mesh` arg.

File: torchtitan/experiments/simple_fsdp/README.md [hunk 8]
```diff
@@ -17,7 +17,7 @@ Some of the features require the updates from PyTorch, with which we are working
 |Meta Initialization| ✅ |
 |Activation Checkpointing| ✅ |
 |Mixed Precision Training| ✅ |
-|Tensor Parallelism| 🚧 |
+|Tensor Parallelism| ✅ |
 |Context Parallelism| ✅ |
 |Pipeline Parallelism| ✅ |
 |Distributed Checkpointing| 🚧 |
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/README.md:23]: I think we can mark Float8 as WIP as well. It seems running with speedup but slower loss converging (there might be bugs). So we need to understand if it's expected or what's the gap.

**Comment by fegin** [comment]
The current DeviceMesh requires users to understand the finest-grained submesh first and use `slicing` and `flatten()` to create the submesh, which is proved hard to use. This PR is a good example. Will add to the DeviceMesh roadmap.

File: torchtitan/experiments/simple_fsdp/simple_fsdp.py [hunk 9]
```diff
@@ -86,10 +159,11 @@ def replicate_compute(self, x):
         # support for FSDP + TP (assuming TP shards the inner-most dim)
         if self.mode == "fully_shard" and x._spec.mesh.ndim == 2:
             dp_placement, tp_placement = x._spec.placements
-            dp_mesh, tp_mesh = self.device_mesh, x._spec.mesh["tp"]
+            # TODO: add x._spec.mesh["tp"] back after DTensor supports submesh slice with different root mesh names
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/simple_fsdp.py]:
```suggestion
            # TODO: remove tp_mesh as an input arg to data_parallel API and use x._spec.mesh["tp"] after DeviceMesh supports slicing a non-root mesh
```

**Review by tianyu-l — APPROVED** [review]
LGTM! Please address final nit before merging.
